src/t8_forest/t8_forest_cxx.cxx

```diff
@@ -1065,14 +1065,95 @@ t8_forest_element_face_normal (t8_forest_t forest, t8_locidx_t ltreeid, const t8
   }
 }
 
-/* Query whether a point lies inside a linear triangle defined by the tree 
- * vertices p_0, p_1, p_2.
+/**
+ * Check if a point lies inside a vertex
+ * 
+ * \param[in] vertex_coords The coordinates of the vertex
+ * \param[in] point         The coordinates of the point to check
+ * \param[in] tolerance     A double > 0 defining the tolerance
+ * \return                  0 if the point is outside, 1 otherwise.  
  */
 static int
-t8_triangle_point_inside (const double p_0[3], const double p_1[3], const double p_2[3], const double point[3],
-                          const double tolerance)
+t8_vertex_point_inside (const double vertex_coords[3], const double point[3], const double tolerance)
+{
+  T8_ASSERT (tolerance > 0);
+  if (t8_vec_dist (vertex_coords, point) > tolerance) {
+    return 0;
+  }
+  return 1;
+}
+
+/**
+ * Check if a point is inside a line that is defined by a starting point \a p_0
+ * and a vector \a vec
+ * 
+ * \param[in] p_0         Starting point of the line
+ * \param[in] vec         Direction of the line (not normalized)
+ * \param[in] point       The coordinates of the point to check
+ * \param[in] tolerance   A double > 0 defining the tolerance
+ * \return                0 if the point is outside, 1 otherwise.  
+ */
+static int
+t8_line_point_inside (const double *p_0, const double *vec, const double *point, const double tolerance)
 {
+  T8_ASSERT (tolerance > 0);
+  double b[3];
+  /* b = p - p_0 */
+  t8_vec_axpyz (p_0, point, b, -1);
+  double x = 0; /* Initialized to prevent compiler warning. */
+  int i;
+  /* So x is the solution to
+  * vx = b.
+  * We can compute it as
+  * x = b[i] / v[i]
+  * if any v[i] is not 0.
+  *
+  * Otherwise the line is degenerated (which should not happen).
+  */
+  for (i = 0; i < 3; ++i) {
+    if (vec[i] != 0) {
+      x = b[i] / vec[i];
+      break; /* found a non-zero coordinate. We can stop now. */
+    }
+  }
+
+  /* If i == 3 here, then v = 0 and hence the line is degenerated. */
+  SC_CHECK_ABORT (i < 3, "Degenerated line element. Both endpoints are the same.");
+
+  if (x < -tolerance || x > 1 + tolerance) {
+    /* x is not an admissible solution. */
+    return 0;
+  }
 
+  /* we can check whether x gives us a solution by
+     * checking whether
+     *  vx = b
+     * is actually true.
+     */
+  double vec_check[3] = { vec[0], vec[1], vec[2] };
+  t8_vec_ax (vec_check, x);
+  if (t8_vec_dist (vec_check, b) > tolerance) {
+    /* Point does not lie on the line. */
+    return 0;
+  }
+  /* The point is on the line. */
+  return 1;
+}
+
+/**
+ * Check if a point is inside of a triangle described by a point \a p_0 and two vectors \a v and \a w. 
+ * 
+ * \param[in] p_0         The first vertex of a triangle
+ * \param[in] v           The vector from p_0 to p_1 (second vertex in the triangle)
```

Comment:
Do we count from zero or one here?
```suggestion
 * \param[in] v           The vector from p_0 to p_1 (first vertex in the triangle)
```

Reply:
No, p_0 is the first vertex in the triangle, p_1 the second and p_2 the third.